builtin/mainmenu/settings/components.lua

```diff
@@ -0,0 +1,378 @@
+						value = math.max(value, setting.min)
+					end
+					if setting.max then
+						value = math.min(value, setting.max)
+					end
+					core.settings:set(setting.name, tostring(value))
+					return true
+				end
+			end,
+		}
+	end
+end
+
+
+make.float = make_field(tonumber, is_valid_number)
+make.int = make_field(function(x) return math.floor(tonumber(x)) end, is_valid_number)
+make.string = make_field(tostring, nil)
+
+
+function make.bool(setting)
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local value = core.settings:get_bool(setting.name, core.is_yes(setting.default))
+			self.changed = tostring(value) ~= setting.default
+
+			local fs = ("checkbox[0,0.25;%s;%s;%s]"):format(
+				setting.name, get_label(setting), tostring(value))
+			return fs, 0.5
+		end,
+
+		on_submit = function(self, fields)
+			if fields[setting.name] == nil then
+				return false
+			end
+
+			core.settings:set_bool(setting.name, core.is_yes(fields[setting.name]))
+			return true
+		end,
+	}
+end
+
+
+function make.enum(setting)
+	return {
+		info_text = setting.comment,
+		setting = setting,
+		max_w = 4.5,
+
+		get_formspec = function(self, avail_w)
+			local value = core.settings:get(setting.name) or setting.default
+			self.changed = value ~= setting.default
+
+			local items = {}
+			for i, option in ipairs(setting.values) do
+				items[i] = core.formspec_escape(option)
+			end
+
+			local selected_idx = table.indexof(setting.values, value)
+			local fs = "label[0,0.1;" .. get_label(setting) .. "]"
+
+			fs = fs .. ("dropdown[0,0.3;%f,0.8;%s;%s;%d]"):format(
+				avail_w, setting.name, table.concat(items, ","), selected_idx, value)
+
+			return fs, 1.1
+		end,
+
+		on_submit = function(self, fields)
+			local old_value = core.settings:get(setting.name) or setting.default
+			local value = fields[setting.name]
+			if value == nil or value == old_value then
+				return false
+			end
+
+			core.settings:set(setting.name, value)
+			return true
+		end,
+	}
+end
+
+
+function make.path(setting)
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local value = core.settings:get(setting.name) or setting.default
+			self.changed = value ~= setting.default
+
+			local fs = ("field[0,0.3;%f,0.8;%s;%s;%s]"):format(
+				avail_w - 3, setting.name, get_label(setting), value)
+			fs = fs .. ("button[%f,0.3;1.5,0.8;%s;%s]"):format(avail_w - 3, "pick_" .. setting.name, fgettext("Browse"))
+			fs = fs .. ("button[%f,0.3;1.5,0.8;%s;%s]"):format(avail_w - 1.5, "set_" .. setting.name, fgettext("Set"))
+
+			return fs, 1.1
+		end,
+
+		on_submit = function(self, fields)
+			local dialog_name = "dlg_path_" .. setting.name
+			if fields["pick_" .. setting.name] then
+				local is_file = setting.type ~= "path"
+				core.show_path_select_dialog(dialog_name,
+					is_file and fgettext_ne("Select file") or fgettext_ne("Select directory"), is_file)
```

Comment:
```suggestion
					is_file and fgettext("Select file") or fgettext("Select directory"), is_file)
```

Reply:
`core.show_path_select_dialog` is not a formspec function so should not use `formspec_escape`

Reply:
Ah, of course